doc/cli_wrapper/shd-cli_wrapper_hld.md

```diff
@@ -0,0 +1,213 @@
+
+* CLI worker decodes the command to get parameters and environment setting, then execute this command based on those settings. In the end, it sends return code and data back to CLI wrapper client
+
+### 1.3.3 Watchdog execution flow
+
+![Command execution flow](images/watchdog_flow.jpg)
+* Default polling time is 60 seconds, watchdog check if the pid in local DB is still running, if not, update local DB.
+
+* Default expired time is 600 seconds (To send terminate signal) and 660 seconds (To send kill signal), watchdog check if CLI worker execution time exceeds expired time, send terminate or kill signal to recycle it.
+
+### 1.3.4 Encoded command format
+
+In order to send all needed information in a request, some encode processes must be applied before sending it to server.
+The following picture shows two examples for encoded requests.
+
+![Encode format](images/encode_format.jpg)
+
+* \__wrapper_calling_cmd__: used to indicate the start of a command
+
+* \__wrapper_env__ : used to indicate the environment setting when executing this command, separate the value by ";", this is a optional field.
+
+### 1.3.5 CLI wrapper service
+
+In compile stage, CLI_wrapper builder put generated wrapped file and original main.py to a backup folder.
+
+CLI wrapper service use these backup files to enable/disable CLI_wrapper feature.
+
+![Encode format](images/compiler_stage.jpg)
+
+* Enable CLI wrapper service
+
+![Encode format](images/wrapper_enable.jpg)
+
+    1. Replace main.py with wrapped main.py
+
+    2. Start CLI wrapper daemon
+
+* Disable CLI wrapper service
+
+![Encode format](images/wrapper_disable.jpg)
+
+    1. Replace main.py with main_impl.py
+
+    2. Stop CLI wrapper daemon
+
+### 1.3.6 Source tree 
+
+Here lists the related source tree of CLI_wrapper
+
+![Encode format](images/source_tree.jpg)
+
+* CLI_wrapper builder : files/build_scripts/cli_wrapper
+* CLI_wrapper client : src/sonic-utilities/cli_wrapper
+* CLI_wrapper server : src/sonic-utilities/cli_wrapper
+* Integrate CLI wrapper builder to sonic-utilities debian: rules/sonic-utilities.mk
+
+# 2. Performance improvements
+
+* The following test results are made in arm platform(dual-core A53 processor):
+
+  * Executing "show vlan brief" command 20 times
+
+     - The following results are compared with and without cli_wrapper.
+
+        ![Encode format](images/test_result_arm_1.jpg)
+
+  * The following is an analysis of the execution time of the "show vlan brief" command
+
+     - The following results are compared with and without cli_wrapper.
+
+       ![Encode format](images/test_result_arm_2.jpg)
+
+* Conclusion
+
+  * Cli_wrapper can improve command execution time when Python code imports more global packages.
+
+  * For less powerful CPUs, cli_wrapper can significantly reduce execution time.
+
+# 3. Open questions
+
+  * TODO:
+
+    * Dynamically generating wrapped files
+

```

Comment:
Please also check the interoperability with GCU feature. Since GCU loads the yang models once for every command, can we think about loading the entire yang only once and then validating?